packages/plugin-data/src/api/data/tree/exportApi.ts

```diff
@@ -0,0 +1,607 @@
+    type: string,
+    saveRefs: boolean,
+    resolveRefs: boolean,
+    fileName: string,
+    sObject: any
+  ): DataPlanPart {
+    let finalFilename = fileName;
+    if (this.config.prefix) {
+      finalFilename = `${this.config.prefix}-${fileName}`;
+    }
+    const dataPlanPart = {
+      sobject: type,
+      saveRefs,
+      resolveRefs,
+      files: [finalFilename],
+    };
+
+    this.writeDataFileSync(fileName, sObject);
+
+    return dataPlanPart;
+  }
+
+  /**
+   * This method is used as a second pass to establish references that couldn't be determined
+   * in the initial pass done by processRecordList. It looks for relationship fields that
+   * contain an id.
+   */
+  private finalApplyRefs(sobjectTree: any): { records: any } {
+    sobjectTree.forEach((record: any) => {
+      Object.keys(record).map((field) => {
+        if (record[field].records) {
+          // These are children
+          this.finalApplyRefs(record[field].records);
+        } else {
+          const objType: string = record.attributes.type;
+
+          if (this.isRelationship(objType, field)) {
+            const id: string = record[field].toString();
+            if (!id.startsWith('@')) {
+              const refTo = this.getRelatedTo(objType, field);
+              const ref: string = this.referenceRegistry[refTo][id];
+
+              if (!ref) {
+                throw new SfdxError(`${objType} reference to ${refTo} (${id}) not found in query results.`);
+              }
+
+              record[field] = `@${ref}`;
+
+              // Setup dependency ordering for later output
+              if (this.objectTypeRegistry[objType].order <= this.objectTypeRegistry[refTo].order) {
+                this.objectTypeRegistry[objType].order = this.objectTypeRegistry[refTo].order + 1;
+                this.objectTypeRegistry[refTo].saveRefs = true;
+                this.objectTypeRegistry[objType].resolveRefs = true;
+              }
+            }
+          }
+        }
+
+        return field;
+      });
+    });
+
+    return { records: sobjectTree };
+  }
+
+  private countRecords(records: AnyArray, count = 0): number {
+    count += records.length;
+    records.forEach((record: any) => {
+      Object.values(record).forEach((val: any) => {
+        if (val?.records) {
+          this.countRecords(val.records, count);
+        }
+      });
+    });
+    return count;
+  }
+
+  private writeDataFileSync(fileName: string, jsonObject: AnyJson): AnyJson {
+    let recordCount = 0;
+    const { outputDir, prefix } = this.config;
+
+    if (prefix) {
+      fileName = `${prefix}-${fileName}`;
+    }
+
+    if (outputDir) {
+      fileName = path.join(outputDir, fileName);
+    }
+
+    if (has(jsonObject, 'records') && isArray(jsonObject.records)) {
+      recordCount = this.countRecords(jsonObject.records);
+    }
+
+    fs.writeFileSync(fileName, JSON.stringify(jsonObject, null, 4));
+
+    // TODO: move this to the command
```

Comment:
let's do it

Reply:
I agree, and I tried but it causes a huge snowball that I thought best to handle in the refactor work items.